### PR TITLE
[X86][GlobalISel] Enable scalar versions of G_UITOFP and G_FPTOUI

### PR DESCRIPTION
**Pull request by e-kud**
Also add tests for G_SITOFP and G_FPTOSI

### CODE REVIEW

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp [hunk 1]
```diff
@@ -497,6 +497,53 @@ X86LegalizerInfo::X86LegalizerInfo(const X86Subtarget &STI,
...
+      .legalIf([=](const LegalityQuery &Query) {
+        return HasAVX512 && typeInSet(0, {s32, s64})(Query) &&
+               typeInSet(1, {s32, s64})(Query);
+      })
+      .customIf([=](const LegalityQuery &Query) {
+        if (HasAVX512)
+          return false;
+        return (HasSSE1 &&
+                (typePairInSet(0, 1, {{s32, s32}, {s32, s16}})(Query) ||
+                 (Is64Bit && typePairInSet(0, 1, {{s32, s64}})(Query)))) ||
```
Comment (RKSimon) [inline llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:517]: For uitofp, SSE has to cheat and zero-extend to i16->i32 / i32->i64 and use sitofp - so we need Is64Bit for i32->f32 as well. We only get real uitofp with AVX512

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp [hunk 2]
```diff
@@ -644,6 +695,112 @@ bool X86LegalizerInfo::legalizeBuildVector(MachineInstr &MI,
...
+  const LLT s64 = LLT::scalar(64);
+
+  // Simply reuse SITOFP when it is possible to widen the type
+  if (SrcTy == s16 || SrcTy == s32) {
+    const LLT WidenTy = LLT::scalar(SrcTy.getScalarSizeInBits() * 2);
+    auto Ext = MIRBuilder.buildZExt(WidenTy, Src);
+    MIRBuilder.buildSITOFP(Dst, Ext);
+    MI.eraseFromParent();
+    return true;
+  }
```
Comment (arsenm) [inline llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:756]: This is reinventing widenScalar on the source

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp [hunk 3]
```diff
@@ -644,6 +695,112 @@ bool X86LegalizerInfo::legalizeBuildVector(MachineInstr &MI,
...
+  if (SrcTy == s64 && DstTy == s32) {
```
Comment (arsenm) [inline llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:757]: This can go in the generic LegalizerHelper

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp [hunk 4]
```diff
@@ -644,6 +695,112 @@ bool X86LegalizerInfo::legalizeBuildVector(MachineInstr &MI,
...
+    auto RoundedHalved = MIRBuilder.buildOr(SrcTy, Halved, LowerBit);
+    auto HalvedFP = MIRBuilder.buildSITOFP(DstTy, RoundedHalved);
+    auto LargeResult = MIRBuilder.buildFAdd(DstTy, HalvedFP, HalvedFP);
+    auto IsLarge = MIRBuilder.buildICmp(CmpInst::Predicate::ICMP_SLT,
+                                        LLT::scalar(1), Src, Zero);
+    MIRBuilder.buildSelect(Dst, IsLarge, LargeResult, SmallResult);
+    MI.eraseFromParent();
+    return true;
+  }
+  if (SrcTy == s64 && DstTy == s64) {
```
Comment (arsenm) [inline llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:776]: This is also generic

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp (hunk 2 above)
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:756]: It is not. Maybe I'm missing something but `customIf` works after `widenScalar`. Given `%1(s32) = G_UITOFP %0(s16)`, it becomes `%1(s32) = G_UITOFP %0(s32)`, and only then we see it in `customIf`. In `legalizeUITOFP` we extend `s32` to `s64` because there it's the only way. What do we have at the end? We extend anything `<s32` to `s64` because `widenScalar` loses the information about the original type. The reinvention allows us only extending  `s16` up to `s32`.

In other words, after `windeScalar` we don't know whether the original type was less than `s32` or it is `s32`and we need either lookup through instructions or extend everything to `s64` to be correct.

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp (hunk 3 above)
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:757]: There is already code in generic `LegalizerHelper`. The code implements `UITOFP` without `SITOFP`. So is your suggestion to add something like `lowerUITOFPAsSITOFP` to `LegalizerHelper`?

Comment (arsenm) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:757]: Yes. It should be a library of different implementations. You can add some heuristic to pick which path to go down in the default

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:756]: > Maybe I'm missing something but customIf works after widenScalar

The order is whatever the rules you explicitly ordered do

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp (hunk 1 above)
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp]: Fixed.

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp (hunk 2 above)
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp]: Thanks! It helped to avoid this ugly `HasAVX512 ? s32 : s16`. But still, I don't get it how to avoid the reinvention. If we widen `s32->s64` before custom lowering, we can't differ `s32->s64` and original `s64`. `s32->s64` should be lowered through `SITOFP`, when `s64` is lowered without bit logic.

Or have you meant that we need to use `widenScalarSrc` here, instead of creating `ZExt` on our own?

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp (hunk 3 above)
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp]: Done

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp (hunk 4 above)
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp]: Done

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 5]
```diff
@@ -6931,6 +6931,77 @@ LegalizerHelper::lowerU64ToF32BitOps(MachineInstr &MI) {
...
+  // Note that + and - are float operations that adjust the implicit leading
+  // one, the bases 2^52 and 2^84 are for illustrative purposes.
+  //
+  // X = 2^52 * 1.0...LowBits
+  // Y = 2^84 * 1.0...HighBits
+  // Scratch = 2^84 * 1.0...HighBits - 2^84 * 1.0 - 2^52 * 1.0
+  //         = - 2^52 * 1.0...HighBits
+  // Result = - 2^52 * 1.0...HighBits + 2^52 * 1.0...LowBits
+  auto TwoP52 = MIRBuilder.buildConstant(S64, UINT64_C(0x4330000000000000));
+  auto TwoP84 = MIRBuilder.buildConstant(S64, UINT64_C(0x4530000000000000));
```
[the remark this answers is not in the document]
Comment (e-kud) [reply llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:6988]: Clang-format doesn't complain. Do you have something specific in mind?

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp [hunk 6]
```diff
@@ -644,6 +691,77 @@ bool X86LegalizerInfo::legalizeBuildVector(MachineInstr &MI,
...
+
+  // Simply reuse FPTOSI when it is possible to widen the type
+  if (DstSizeInBits <= 32) {
+    auto Casted = MIRBuilder.buildFPTOSI(DstTy == s32 ? s64 : s32, Src);
+    MIRBuilder.buildTrunc(Dst, Casted);
+    MI.eraseFromParent();
+    return true;
+  }
+
+  if (DstTy == s64) {
```
[the remark this answers is not in the document]
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:711]: I don't think so, it relies on the fact that X86 version of `FPTOSI` returns `80000000_00000000H` in case of failed conversion.

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp [hunk 7]
```diff
@@ -644,6 +691,77 @@ bool X86LegalizerInfo::legalizeBuildVector(MachineInstr &MI,
...
+  }
+
+  if (SrcTy == s64 && DstTy == s32)
+    return Helper.lowerU64ToF32WithSITOFP(MI) !=
+           LegalizerHelper::LegalizeResult::UnableToLegalize;
+
+  if (SrcTy == s64 && DstTy == s64)
+    // TODO: rewrite with vector shuffles when supported.
+    return Helper.lowerU64ToF64BitFloatOps(MI) !=
+           LegalizerHelper::LegalizeResult::UnableToLegalize;
```
[the remark this answers is not in the document]
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:760]: Oh, that's great! If we can lower `s64` as a default lower action, we can finally use `widenScalar` before `customIf` and get rid of manual widening.
The problem here is that there is already the implementation for `s64->s32` that is not efficient in terms of `X86` as the more efficient is with `SITOFP`. However I can't find a target using `lower` for `UITOFP`. So I can replace the lowering that is more convenient for `X86`. However in future we either need a hook or some parameters passed to the Helper.

[the remark this answers is not in the document]
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp]: Unfortunately no, we still don't distinguish between smaller types promoted to `s32` and original `s32` because after widening `s8->s32` we start rule iteration over and have no clue whether we should widen type to `s64` or not.

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 8]
```diff
@@ -7108,6 +7108,77 @@ LegalizerHelper::lowerU64ToF32BitOps(MachineInstr &MI) {
...
+  auto SmallResult = MIRBuilder.buildSITOFP(S32, Src);
+  // Result if Src >= INT_MAX
+  auto Halved = MIRBuilder.buildLShr(S64, Src, One);
+  auto LowerBit = MIRBuilder.buildAnd(S64, Src, One);
+  auto RoundedHalved = MIRBuilder.buildOr(S64, Halved, LowerBit);
+  auto HalvedFP = MIRBuilder.buildSITOFP(S32, RoundedHalved);
+  auto LargeResult = MIRBuilder.buildFAdd(S32, HalvedFP, HalvedFP);
+  // Choose
+  auto IsLarge =
+      MIRBuilder.buildICmp(CmpInst::Predicate::ICMP_SLT, S1, Src, Zero);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:7138]: IsLarge = slt x, 0?

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 9]
```diff
@@ -7108,6 +7108,77 @@ LegalizerHelper::lowerU64ToF32BitOps(MachineInstr &MI) {
...
+  auto Zero = MIRBuilder.buildConstant(S64, 0);
+  // Result if Src < INT_MAX
+  auto SmallResult = MIRBuilder.buildSITOFP(S32, Src);
+  // Result if Src >= INT_MAX
+  auto Halved = MIRBuilder.buildLShr(S64, Src, One);
+  auto LowerBit = MIRBuilder.buildAnd(S64, Src, One);
+  auto RoundedHalved = MIRBuilder.buildOr(S64, Halved, LowerBit);
+  auto HalvedFP = MIRBuilder.buildSITOFP(S32, RoundedHalved);
+  auto LargeResult = MIRBuilder.buildFAdd(S32, HalvedFP, HalvedFP);
+  // Choose
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:7136]: Incomplete comment?

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:717]: You can't really directly rely on known target instruction behavior and still use the regular G_* opcodes. Other optimizations could do something assuming the poison behavior 

Also add that as a comment somewhere?

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp [hunk 10]
```diff
@@ -498,6 +498,54 @@ X86LegalizerInfo::X86LegalizerInfo(const X86Subtarget &STI,
...
+               typeInSet(1, {s32, s64})(Query);
+      })
+      .lowerIf([=](const LegalityQuery &Query) {
+        // Lower conversions from s64
+        return !HasAVX512 &&
+               ((HasSSE1 && typeIs(0, s32)(Query)) ||
+                (HasSSE2 && typeIs(0, s64)(Query))) &&
+               (Is64Bit && typeIs(1, s64)(Query));
+      })
+      .customIf([=](const LegalityQuery &Query) {
```
Comment (arsenm) [inline llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:520]: Move customIf before lowerIf and after legalIf

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp (hunk 8 above)
Comment (e-kud) [reply llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:7139]: Yes because `x` is unsigned. Or do we want to use ult and INT_MAX to make it more explicit?

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp (hunk 9 above)
Comment (e-kud) [reply llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp]: Actually no but I added a few lines considering the comment above about `slt x, 0`

**Comment by e-kud** [comment]
> You can't really directly rely on known target instruction behavior and still use the regular G_* opcodes. Other optimizations could do something assuming the poison behavior

Yes, this is true, thank you for noticing this! Updated version uses the lowering from the helper. The problem with `X86cvtts2Int` is that there are patterns only for vector form in contrast to `fp_to_sint`. We currently struggle with it in GlobalISel.

Probably I could've hacked it by `GINodeEquiv<G_CVTTS2INT, fp_to_sint>` but I'm inclined to think that target specific opcodes should be mapped to target specific nodes.

**Review by arsenm — APPROVED** [review]

**Comment by arsenm** [comment]
> Yes, this is true, thank you for noticing this! Updated version uses the lowering from the helper. The problem with `X86cvtts2Int` is that there are patterns only for vector form in contrast to `fp_to_sint`. We currently struggle with it in GlobalISel.

I think this is fine as-is to make progress, but I think this needs to be cleaned up in a follow up. I would suggest either emitting explicit clamping code and having an x86 combine match the instruction behavior, or introduce another G_* node with the clamping behavior and let other targets lower that with the clamp code

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp [hunk 11]
```diff
@@ -498,6 +498,64 @@ X86LegalizerInfo::X86LegalizerInfo(const X86Subtarget &STI,
...
+      .legalIf([=](const LegalityQuery &Query) {
+        return HasAVX512 && typeInSet(0, {s32, s64})(Query) &&
+               typeInSet(1, {s32, s64})(Query);
+      })
+      .customIf([=](const LegalityQuery &Query) {
+        return !HasAVX512 &&
+               ((HasSSE1 && typeIs(0, s32)(Query)) ||
+                (HasSSE2 && typeIs(0, s64)(Query))) &&
+               (scalarNarrowerThan(1, 32)(Query) ||
+                (Is64Bit && typeIs(1, s32)(Query)));
```
Comment (RKSimon) [inline llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:518]: Could you simplify this to `scalarNarrowerThan(1, Is64Bit ? 64 : 32)`?

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp [hunk 12]
```diff
@@ -498,6 +498,64 @@ X86LegalizerInfo::X86LegalizerInfo(const X86Subtarget &STI,
...
+      .legalIf([=](const LegalityQuery &Query) {
+        return HasAVX512 && typeInSet(0, {s32, s64})(Query) &&
+               typeInSet(1, {s32, s64})(Query);
+      })
+      .customIf([=](const LegalityQuery &Query) {
+        return !HasAVX512 &&
+               ((HasSSE1 && typeIs(1, s32)(Query)) ||
+                (HasSSE2 && typeIs(1, s64)(Query))) &&
+               (scalarNarrowerThan(0, 32)(Query) ||
+                (Is64Bit && typeIs(0, s32)(Query)));
```
Comment (RKSimon) [inline llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:542]: `scalarNarrowerThan(0, Is64Bit ? 64 : 32)`?

File: llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp (hunk 11 above)
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp]: Sure, thanks!

File: llvm/test/CodeGen/X86/isel-fp-to-int.ll [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+; SDAG-AVX512-NEXT:    # kill: def $al killed $al killed $eax
+; SDAG-AVX512-NEXT:    retq
+;
+; GISEL-AVX512-LABEL: test_float_to_ui8:
+; GISEL-AVX512:       # %bb.0: # %entry
+; GISEL-AVX512-NEXT:    vcvttss2usi %xmm0, %eax
+; GISEL-AVX512-NEXT:    # kill: def $al killed $al killed $eax
+; GISEL-AVX512-NEXT:    retq
+entry:
+  %conv = fptoui float %x to i8
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/isel-fp-to-int.ll:201]: Looks like we're missing fold to prefer to use fp_to_sint when legal

**Review by RKSimon — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/X86/isel-fp-to-int.ll (hunk 13 above)
Comment (e-kud) [reply llvm/test/CodeGen/X86/isel-fp-to-int.ll:201]: Are there any reasons to prefer `vcvttss2si` over `vcvttss2usi` in case of AVX512?

Comment (topperc) [reply llvm/test/CodeGen/X86/isel-fp-to-int.ll:201]: vcvttss2si on 128/256-bit vectors has a shorter encoding when it can use VEX instead of EVEX.

**Comment by e-kud** [comment]
> I think this is fine as-is to make progress, but I think this needs to be cleaned up in a follow up. I would suggest either emitting explicit clamping code and having an x86 combine match the instruction behavior, or introduce another G_* node with the clamping behavior and let other targets lower that with the clamp code

@arsenm I'm not sure if a new opcode is a good idea. x86 version clamps everything, that is out of bounds, to `80000000_00000000H`. So to make it work, we need to introduce an opcode for generic clamping to the specified value or values.

**Comment by arsenm** [comment]
> @arsenm I'm not sure if a new opcode is a good idea. x86 version clamps everything, that is out of bounds, to `80000000_00000000H`. So to make it work, we need to introduce an opcode for generic clamping to the specified value or values.

It is perfectly fine to define a new opcode with whatever the clamping behavior exists. I would suggest it's easier to just pattern match the explicit clamp code though
